specifyweb/frontend/js_src/lib/components/Statistics/hooks.tsx

```diff
@@ -0,0 +1,424 @@
+      unknownCategories.includes(categoryToFetch as keyof typeof urlSpec)
+    );
+  }, [layout]);
+}
+
+export function statsToTsv(
+  layout: IR<StatLayout | undefined>,
+  sourceIndex: number,
+  layoutPageIndex: number
+): { readonly statsTsv: string; readonly nameSpec: string } | undefined {
+  if (Object.values(layout).some((layouts) => layouts === undefined))
+    return undefined;
+  const headers = [
+    statsText.source(),
+    statsText.pageName(),
+    statsText.categoryName(),
+    statsText.itemName(),
+    statsText.itemValue(),
+  ];
+  const statItems: WritableArray<WritableArray<number | string>> = [];
+  let nameSpec = '';
+  Object.entries(layout as IR<StatLayout>).forEach(
+    ([sourceName, layouts], layoutSourceIndex) =>
+      layouts.forEach((layout, pageIndex) => {
+        if (layout === undefined) return;
+        const layoutLabel = layout.label === undefined ? '' : layout.label;
+        layout.categories.forEach((category) => {
+          if (category === undefined) return;
+          const categoryLabel =
+            category.label === undefined ? '' : category.label;
+          if (category.items === undefined) return;
+          category.items.forEach(({ itemLabel, itemValue }) => {
+            if (itemValue === undefined) return;
+            const newItemLabel = itemLabel === undefined ? '' : itemLabel;
+            if (
+              layoutSourceIndex === sourceIndex &&
+              pageIndex === layoutPageIndex
+            ) {
+              nameSpec = `${sourceName} ${layoutLabel}`;
+              statItems.push([
+                sourceName,
+                layoutLabel,
+                categoryLabel,
+                newItemLabel,
+                itemValue.toString(),
+              ]);
+            }
+          });
+        });
+      })
+  );
+  return {
+    statsTsv: [headers, ...statItems].map((line) => line.join('\t')).join('\n'),
+    nameSpec,
+  };
+}
+
+export function useStatValueLoad<
+  PROMISE_TYPE extends string | number | undefined
+>(
+  statValue: string | number | undefined,
+  promiseGenerator: () => Promise<PROMISE_TYPE>,
+  handleValueLoad: ((value: number | string) => void) | undefined
+) {
+  const shouldFetch =
+    statValue === undefined && typeof handleValueLoad === 'function';
+  React.useEffect(() => {
+    if (!shouldFetch) return undefined;
+    let destructorCalled = false;
+    promiseGenerator().then((value) => {
+      if (value === undefined || destructorCalled) return;
+      handleValueLoad?.(value);
+    });
+    return (): void => {
+      destructorCalled = true;
+    };
+  }, [promiseGenerator, statValue, handleValueLoad]);
+}
+
+export function useUnknownCategory(
+  backEndResponse: BackendStatsResult | undefined,
+  handleChange: (
+    newCategories: (
+      oldCategory: StatLayout[number]['categories']
+    ) => StatLayout[number]['categories']
+  ) => void,
+  statsSpec: StatsSpec
+) {
+  React.useLayoutEffect(() => {
```

Comment:
I think so - for add stat dialog, unknown categories are fetched when the page is loaded, but the filter condition for default stats to add skips over categories with undefined items - so the unknown categories will be inserted into the default layout after the add stat dialog opens (so they will see category appearing after the page opens). We can have it display "loading" if you think that would be better. useLayoutEffect ensures that the default layout will already have the unknown categories since it is called before the page is painted by the browser.